DietPi-LetsEncrypt | Error handling + further rework

MichaIng opened the pull request:
Via Forum: https://dietpi.com/phpbb/viewtopic.php?f=11&t=1909
+ DietPi-LetsEncrypt | Skip any webserver config edit, if CertBot execution returns error: https://dietpi.com/phpbb/viewtopic.php?f=11&t=1909
+ DietPi-LetsEncrypt | Lighttpd SSL config upgrade according to Mozilla SSL generator: https://mozilla.github.io/server-side-tls/ssl-config-generator/?server=lighttpd-1.4.35&openssl=1.0.1t&hsts=yes&profile=intermediate
+ DietPi-LetsEncrypt | Minor code and error handling improvements, revert path variable naming scheme
   - `ps aux | grep` vs `pgrep`: https://github.com/Fourdee/DietPi/issues/1510#issuecomment-385370492
- [x] Added changelog entries
________________
<s>**WIP: testing required**</s>
**Testing and fixing done, ready to merge!**
- [x] Stretch + Apache until certbot cmd
- [x] Stretch + Nginx until certbot cmd
- [x] Stretch + Lighttpd until certbot cmd
- [x] Stretch + Minio until certbot cmd
- [x] Stretch + Apache until end (add self-signed certs for webserver links/tasks)
- [x] Stretch + Nginx until end (add self-signed certs for webserver links/tasks)
- [x] Stretch + Lighttpd until combined.pem error handler
- [x] Stretch + Lighttpd until end (add self-signed certs for webserver links/tasks)
  - 🈯️ Error with `(` inside pattern. Treated specially with ext. regex, expecting `)`, `\(` just works for `grep`, but `sed` somehow needs additional escaping, which then fails `grep` check. Confusing as with direct call `grep` and `sed` need/understand the exact same level of escaping. Sub sub script/function seems to mix things up... Perhaps we need to remove ext. regex again from `G_CONFIG_INJECT`. **€: Recheck success!**
  - 🈯️ `G_CONFIG_INJECT` pattern as well: `\\t` double escaping or better direct tab symbol needed. **€: Recheck success!**
  - 🈯️ `´´` special symbols inside `cat` leads to execution of `openssl` command. **€: Recheck success!**
  - 🈯️ Edit `sed` commands to just apply on first match! **€: Recheck success!**
  - 🈯️ `sed -Ei` is needed, `sed -iE` will create new file with `E` added to file name. **€: Recheck success!**
- [x] Stretch + Minio until end (add self-signed certs for webserver links/tasks)
- [x] Jessie + Apache until end (add self-signed certs for webserver links/tasks)
  - 🈯️ pgrep `-i` option doesn't exist on Jessie **€: Recheck success!**
- [x] Jessie + Nginx until end (add self-signed certs for webserver links/tasks)
  - 🈯️ G_CONFIG_INJECT fix sed adds instead of replaces accidentally **€: Recheck success!**
- [x] Jessie + Lighttpd until end (add self-signed certs for webserver links/tasks)
  - 🈯️ Just add `2>> logfile` after cat, to just log errors while allowing to forward file contents to combined.pem **€: Recheck success!**
- [x] Jessie + Minio until end (add self-signed certs for webserver links/tasks)
________________
- <s> [ ] Add CertBot from backports to re-enable TLS-SNI authentication method on Stretch?</s> **Adding to separate PR**

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Skip any webserver config edit, if CertBot execution returns error: https://dietpi.com/phpbb/viewtopic.php?f=11&t=1909
+ DietPi-LetsEncrypt | Lighttpd SSL config upgrade according to Mozilla SSL generator: https://mozilla.github.io/server-side-tls/ssl-config-generator/?server=lighttpd-1.4.35&openssl=1.0.1t&hsts=yes&profile=intermediate
+ DietPi-LetsEncrypt | Minor code and error handling improvements, revert path variable naming scheme

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Skip any webserver config edit, if CertBot execution returns error: https://dietpi.com/phpbb/viewtopic.php?f=11&t=1909
+ DietPi-LetsEncrypt | Lighttpd SSL config upgrade according to Mozilla SSL generator: https://mozilla.github.io/server-side-tls/ssl-config-generator/?server=lighttpd-1.4.35&openssl=1.0.1t&hsts=yes&profile=intermediate
+ DietPi-LetsEncrypt | Minor code and error handling improvements, revert path variable naming scheme

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Syntax fix (whoopsie) + replace "ps aux | grep" by "pgrep"

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Syntax fix: Exit code "$?" will be overwritten by check, thus needs to be assigned to variable before check and further use.

Fourdee commented on dietpi/dietpi-letsencrypt:
```diff
@@ -122,43 +77,79 @@ _EOF_
 			# - Create combined key
-			cd /etc/letsencrypt/live/"$LETSENCRYPT_DOMAIN"
-			cat privkey.pem cert.pem > combined.pem
+			G_RUN_CMD cat $fp_cert_dir/privkey.pem $fp_cert_dir/cert.pem > $fp_cert_dir/combined.pem
```
Redirects and pipes are not supported with ```G_RUN_CMD``` ☹️ 
```
root@DietPi:~# G_RUN_CMD cat /etc/network/interfaces > test
root@DietPi:~# cat test
[  OK  ] Root access verified.
[  OK  ] cat /etc/network/interfaces
```

MichaIng replied:
Indeed, thanks for hint, will make commit later, also to add raw IP test.
€: Done, added alternative error handling, as this is critical.

Fourdee commented on dietpi/dietpi-letsencrypt:
```diff
@@ -193,18 +183,17 @@ _EOF_
 		#------------------------------------------------------------------------------------------------------
 		# Nginx
-		elif (( $(ps aux | grep -ci -m1 '[n]ginx') )); then
+		elif pgrep -i '[n]ginx' &> /dev/null; then
```
Interesting, seems ```pgrep``` automatically removes the process which checks for it, so its not added to the count:
Starting with ```[n]``` is not then needed, but no harm 😃 
```
root@DietPi:~# pgrep -i bash
2318
2402
root@DietPi:~# pgrep -i [b]ash
2318
2402
root@DietPi:~# ps aux | grep bash
root      1002  0.0  0.0   4308   580 pts/0    S+   15:19   0:00 grep --color=auto bash
root      2318  0.0  0.1   6132  4036 ?        Ss   13:36   0:00 -bash
root      2402  0.0  0.1   7144  5320 pts/0    Ss   13:37   0:00 -bash
root@DietPi:~# ps aux | grep [b]ash
root      2318  0.0  0.1   6132  4036 ?        Ss   13:36   0:00 -bash
root      2402  0.0  0.1   7144  5320 pts/0    Ss   13:37   0:00 -bash
```

MichaIng replied:
😆 Jep thought the same, just took `-i` and `[n]` from the old code.
€: Removed [ ]

Fourdee reviewed (changes requested):
G_RUN_CMD changed required, aside from that, looks good 👍

Fourdee commented:
@MichaIng 

> WIP: testing required

Not the easiest thing to test Certbot, hehe 👍

MichaIng commented:
@Fourdee 
Jep always a problem. I will add some fix + raw IP as domain check and do test with all webservers until certbot execution, then locally allow errors and redo to test webserver adjustments.

€: Hmm, just thought, if I'm already touching this, shall we add backports certbot to Stretch to re-allow sni-tls authentication, including `--apache` and `--nginx` options again? Backports are now some versions forward, fixing the pre 0.21 security whole that broke these: https://packages.debian.org/de/stretch-backports/certbot

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Check for valid domain name before executing CertBot
+ DietPi-LetsEncrypt | Code fix and enhancements

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Lighttpd fix

MichaIng committed:
v6.8

+ DietPi-Globals | G_CONFIG_INJECT: Fix "sed" usage to correctly handle options and only apply to first match
+ DietPi-Globals | G_CONFIG_INJECT: Do not create file, if not existent, in case of wrong assume file exists and otherwise wrong variable/usage

MichaIng committed:
v6.8

+ DietPi-LetsEncrypt | Re-enable Jessie support by removing unsupported `-i` option from `pgrep`
+ DietPi-LetsEncrypt | Redirect just errors from `cat` to logfile, to allow merging of combined.pem

MichaIng committed:
v6.8

+ DietPi-Globals | G_CONFIG_INJECT: Whoopsie: Existing settings/comments should be replaced, not extended

MichaIng committed:
v6.8

typo

MichaIng commented:
@Fourdee 
- Testing done. Verified functionality of config adjustments via self-signed certificates and manually ignoring/skipping certbot binary execution of all web browsers on Jessie and Stretch. Also cron job and renewal systemd additions tested.
- I will merge, as there is some urgent G_CONFIG_INJECT fix included that can mess up the testing systems.
- <s>I will open separate PR for stretch-backports certbot, where we could add other ideas according to the above.</s> **€: Ah, TLS-SNI is still blocked by LetsEncrypt, the backports version of CertBot just let's --apache and --nginx use HTTP-01 acme challenge. Thus there is not difference (in security or something) to our current solution, using webroot authentication, at least from my point of view.**
